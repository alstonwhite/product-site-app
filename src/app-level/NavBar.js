import React, { useState } from 'react';
import { Link } from "react-router-dom";
import { useSelector } from "react-redux";

const NavBar = ({sections}) => {

  const cart = useSelector(state => state.cart); 
  let cartItems = cart ? cart.reduce((sum, cur) => sum + cur.quantity, 0) : 0;
  let cartValue = cart ? cart.reduce((sum, cur) => sum + cur.price * cur.quantity, 0) : 0;

  const [open, setOpen] = useState(false); 

  return (
    <div className="nav">
      <button 
        aria-expanded={open} 
        aria-controls="nav-list"
        onClick={()=>setOpen(!open)}  
      >
        <span class="open">☰</span>
        <span class="close">×</span>
        Menu
      </button>
      <Link to="/cart" className = "nav__cart-content btn">
        <div className = "nav__cart-items">{`Cart: ${cartItems} items`}</div>
        <div className = "nav__cart-value">{` $${cartValue}`}</div>
      </Link>
      <ul>
        <li>
          <Link to="/" className = "nav__item">All Products</Link>
        </li>
        <li>
          <Link to="/category/product-group-1" className = "nav__item">Product Group 1</Link>
        </li>
        <li>
          <Link to="/category/product-group-2" className = "nav__item">Product Group 2</Link>
        </li>
        <li>
          <Link to="/category/product-group-3" className = "nav__item">Product Group 3</Link>
        </li>
        {sections && sections.map(section => (
          <li>
            <Link
                to={"/category/"+section.slug}
                className = "nav__item"
            >
              {section.sectionTitle}
            </Link>
          </li>
        ))}
      </ul>
    </div>
    );
  };
  
  export default NavBar;