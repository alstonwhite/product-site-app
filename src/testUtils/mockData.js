export const testSections = 
    [
        {
            id: "0",
            slug: "section1",
            sectionTitle: "Section 1",
            active: true
        },
        {
            id: "1",
            slug: "section2",
            sectionTitle: "Section 2",
            active: true
        }
    ]

export const testProducts = 
    [
        {
            id: "0",
            image1: {fields: {file: {url: "1"}}},
            image2: {fields: {file: {url: "2"}}},
            image3: {fields: {file: {url: "3"}}},
            price: 5,
            title: "Product 1",
            description: "Description",
            productSection: {fields: {slug: "section1"}},
            active: true
        },
        {
            id: "1",
            image1: {fields: {file: {url: "1"}}},
            image2: {fields: {file: {url: "2"}}},
            image3: {fields: {file: {url: "3"}}},
            price: 5,
            title: "Product 2",
            description: "Description",
            productSection: {fields: {slug: "section1"}},
            active: true
        },
        {
            id: "2",
            image1: {fields: {file: {url: "1"}}},
            image2: {fields: {file: {url: "2"}}},
            image3: {fields: {file: {url: "3"}}},
            price: 5,
            title: "Product 3",
            description: "Description",
            productSection: {fields: {slug: "section2"}},
            active: true
        }
    ]

export const testProduct = 
    {
        id: 0,
        image1: {fields: {file: {url: "1"}}},
        image2: {fields: {file: {url: "2"}}},
        image3: {fields: {file: {url: "3"}}},
        price: 5,
        title: "Product 1",
        description: "Description",
        productSection: {fields: {slug: "section1"}},
        active: true
    }

export const testCartProduct = 
[{
    id: 0,
    image1: {fields: {file: {url: "1"}}},
    image2: {fields: {file: {url: "2"}}},
    image3: {fields: {file: {url: "3"}}},
    price: 5,
    title: "Product 1",
    description: "Description",
    productSection: {fields: {slug: "section1"}},
    active: true,
    quantity: 1
}]
