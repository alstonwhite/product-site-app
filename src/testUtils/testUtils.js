import React from "react";
import { Router, Route } from "react-router-dom";
import { render } from "@testing-library/react";
import { createMemoryHistory } from "history";

export const renderWithRouter = (ui, 
    {
        path="/",
        route="/",
        history = createMemoryHistory({ initialEntries: [route] }) 
    }={}) => ( 
    render( 
        <Router history={history}> 
            <Route path={path} children={ui} /> 
        </Router> 
    ) 
)