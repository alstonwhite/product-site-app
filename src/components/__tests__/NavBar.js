import React from 'react'
import { cleanup, screen } from '@testing-library/react'
import NavBar from '../NavBar'
import { testSections } from '../../testUtils/mockData'
import { renderWithRouter } from '../../testUtils/testUtils'


afterEach(cleanup)

jest.mock('react-redux', () => {
    return {
    ...jest.requireActual('react-redux'),
    // cart initializes with item 'testCartProduct' -- hardcoded bc import out of scope 
    useSelector: () => 
    [{
        id: 0,
        image1: {fields: {file: {url: "1"}}},
        image2: {fields: {file: {url: "2"}}},
        image3: {fields: {file: {url: "3"}}},
        price: 5,
        title: "Product 1",
        description: "Description",
        productSection: {fields: {slug: "section1"}},
        active: true,
        quantity: 1
    }],
    useDispatch: jest.fn()
    }
});

test('displays cart quantity and calculated cart value', () => {
    renderWithRouter(<NavBar sections={testSections}/>)
  
    expect(screen.getByText('Cart: 1 item')).toBeInTheDocument()
    expect(screen.getByText('$5')).toBeInTheDocument()
})

test('renders proper number of sections', () => {
    renderWithRouter(<NavBar sections={testSections}/>)
    
    expect(screen.getAllByTestId('nav-section')).toHaveLength(testSections.length)
})

test('snapshot', () => {
    const {container} = renderWithRouter(<NavBar sections={testSections}/>)
    expect(container.firstChild).toMatchSnapshot()
})