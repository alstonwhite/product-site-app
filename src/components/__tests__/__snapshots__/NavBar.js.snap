// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="nav"
>
  <button
    aria-controls="nav-list"
    aria-expanded="false"
    class="nav-hamburger btn"
  >
    <span
      class="open"
    >
      â˜° Menu
    </span>
    <span
      class="close"
    >
      x Menu
    </span>
  </button>
  <a
    class="nav__cart-content btn"
    href="/cart"
  >
    <div
      class="nav__cart-items"
    >
      Cart: 1 item
    </div>
    <div
      class="nav__cart-value"
    >
       $5
    </div>
  </a>
  <ul>
    <li>
      <a
        class="nav__item"
        href="/"
      >
        All Products
      </a>
    </li>
    <li
      data-testid="nav-section"
    >
      <a
        class="nav__item"
        href="/category/section1"
      >
        Section 1
      </a>
    </li>
    <li
      data-testid="nav-section"
    >
      <a
        class="nav__item"
        href="/category/section2"
      >
        Section 2
      </a>
    </li>
  </ul>
</div>
`;
